distributed/scheduler.py

```diff
@@ -4245,167 +4249,165 @@ async def add_nanny(self) -> dict[str, Any]:
         }
         return msg
 
-    def update_graph_hlg(
+    def update_graph(
         self,
-        client=None,
-        hlg=None,
-        keys=None,
-        dependencies=None,
-        restrictions=None,
-        priority=None,
-        loose_restrictions=None,
-        resources=None,
-        submitting_task=None,
-        retries=None,
-        user_priority=0,
-        actors=None,
-        fifo_timeout=0,
-        code=None,
-    ):
-        unpacked_graph = HighLevelGraph.__dask_distributed_unpack__(hlg)
-        dsk = unpacked_graph["dsk"]
-        dependencies = unpacked_graph["deps"]
-        annotations = unpacked_graph["annotations"]
-
-        # Remove any self-dependencies (happens on test_publish_bag() and others)
-        for k, v in dependencies.items():
-            deps = set(v)
-            if k in deps:
-                deps.remove(k)
-            dependencies[k] = deps
+        client: str,
+        graph_header: dict,
+        graph_frames: list[bytes],
+        keys: list[str],
+        internal_priority: dict[str, int] | None,
+        submitting_task: str | None,
+        user_priority: int | dict[str, int] = 0,
+        actors: bool | list[str] | None = None,
+        fifo_timeout: float = 0.0,
+        code: str | None = None,
+        annotations: dict | None = None,
+        stimulus_id: str | None = None,
+    ) -> None:
+        start = time()
+        try:
+            # TODO: deserialization + materialization should be offloaded
+            from distributed.protocol import deserialize
 
-        if priority is None:
-            # Removing all non-local keys before calling order()
-            dsk_keys = set(dsk)  # intersection() of sets is much faster than dict_keys
-            stripped_deps = {
-                k: v.intersection(dsk_keys)
-                for k, v in dependencies.items()
-                if k in dsk_keys
-            }
-            priority = dask.order.order(dsk, dependencies=stripped_deps)
+            graph = deserialize(graph_header, graph_frames).data
+        except Exception as e:
+            err = error_message(e)
+            for key in keys:
+                self.report(
+                    {
+                        "op": "task-erred",
+                        "key": key,
+                        "exception": err["exception"],
+                        "traceback": err["traceback"],
+                    }
+                )
 
-        return self.update_graph(
-            client,
+            return
```

Comment:
@rjzamora @jacobtomlinson Re: breaking changes discussed in https://github.com/dask/dask/pull/9988

This is the place where I anticipate 9 out of 10 failures to occur. Right now we're just passing along the original exception but we could also wrap this in an exception like

> Error during deserialization of the task graph. This frequently occurs if the scheduler and Client have different environments. For more information, see https://docs.dask.org/en/stable/deployment-considerations.html#consistent-software-environments

It could also break in other places below. Specifically during materialization or annotation parsing but there are more things that could go wrong with that and I do not want to slap this very specific message on all exceptions but I'm open for suggestions

Reply:
This idea seems like a pretty-good idea to me.